src/main/java/io/cryostat/net/web/http/api/beta/graph/EnvironmentNodesFetcher.java

```diff
@@ -38,68 +38,75 @@
 package io.cryostat.net.web.http.api.beta.graph;
 
 import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Objects;
+import java.util.Set;
 import java.util.function.Function;
 
 import javax.inject.Inject;
 
 import io.cryostat.platform.discovery.AbstractNode;
 import io.cryostat.platform.discovery.EnvironmentNode;
-import io.cryostat.platform.discovery.TargetNode;
 
 import graphql.schema.DataFetcher;
 import graphql.schema.DataFetchingEnvironment;
-import graphql.schema.DataFetchingEnvironmentImpl;
 
 class EnvironmentNodesFetcher implements DataFetcher<List<EnvironmentNode>> {
 
     private final RootNodeFetcher rootNodeFetcher;
-    private final EnvironmentNodeRecurseFetcher recurseFetcher;
 
     @Inject
-    EnvironmentNodesFetcher(
-            RootNodeFetcher rootNodeFetcher, EnvironmentNodeRecurseFetcher recurseFetcher) {
+    EnvironmentNodesFetcher(RootNodeFetcher rootNodeFetcher) {
         this.rootNodeFetcher = rootNodeFetcher;
-        this.recurseFetcher = recurseFetcher;
     }
 
     @Override
     public List<EnvironmentNode> get(DataFetchingEnvironment environment) throws Exception {
         FilterInput filter = FilterInput.from(environment);
         EnvironmentNode root = rootNodeFetcher.get(environment);
-        List<EnvironmentNode> nodes =
-                recurseFetcher.get(
-                        DataFetchingEnvironmentImpl.newDataFetchingEnvironment(environment)
-                                .source(root)
-                                .build());
+        Set<EnvironmentNode> nodes = flattenEnvNodes(root);
 
         if (filter.contains(FilterInput.Key.NAME)) {
             String nodeName = filter.get(FilterInput.Key.NAME);
-            nodes = recurse(root, n -> Objects.equals(n.getName(), nodeName));
+            nodes = filter(nodes, n -> Objects.equals(n.getName(), nodeName));
         }
 
         if (filter.contains(FilterInput.Key.NODE_TYPE)) {
             String nodeType = filter.get(FilterInput.Key.NODE_TYPE);
-            nodes = recurse(root, n -> n.getNodeType().getKind().equalsIgnoreCase(nodeType));
+            nodes = filter(nodes, n -> Objects.equals(n.getNodeType().getKind(), nodeType));
         }
-        return nodes;
+        return new ArrayList<>(nodes);
     }
 
-    List<EnvironmentNode> recurse(
-            EnvironmentNode node, Function<EnvironmentNode, Boolean> matcher) {
-        List<EnvironmentNode> result = new ArrayList<>();
+    Set<EnvironmentNode> flattenEnvNodes(EnvironmentNode root) {
+        return new HashSet<>(recurse(root, e -> e instanceof EnvironmentNode));
```

Comment:
Is the switch to `HashSet` from `ArrayList` due to performance reasons? I guess in this case, the order doesn't matter and nodes are unique so `HashSet` makes more sense?

Reply:
Right, order doesn't matter/is undefined for this query. It isn't so much a performance reason as a cheap way to de-duplicate the collection. There shouldn't really be duplicates anyway if the tree is truly a tree and not a general graph (with potential cycles), but this is cheap insurance against that possibility.